Wikipedia/Code/WMFAppViewController.m

```diff
@@ -865,8 +866,8 @@ - (void)resumeApp:(dispatch_block_t)completion {
                            completion:^(BOOL didProcess) {
                                done();
                            }];
-        } else if ([self shouldShowLastReadArticleOnLaunch]) {
-            [self showLastReadArticleAnimated:NO];
+        } else if (NSUserDefaults.wmf.shouldRestoreNavigationStackOnResume) {
+            [self.navigationStateController restoreNavigationStateFor:self.navigationController in:self.dataStore.viewContext];
```

Comment:
Since this is asynchronous, the explore feed is showing before the navigation stack is restored. Could this either be made synchronous or could the splash screen remain up until this finishes?

Also, is the login call in the state controller necessary if it's happening again in finishResumingApp?

Reply:
`finishResumingApp` is happening after restoring so there's no guarantee that the user gets logged in before we attempt to show Account VC

Reply:
It seems the Settings VC and Account VC should be able to be shown with ambiguous login state (spinner or otherwise), then respond to login state changes, either by showing the correct info or an error that they aren't actually logged in. This way the app launch and state restoration isn't dependent on a network call. I've noticed this in the past with the app, if I navigate quickly enough to settings, it shows me as logged out even though the app eventually logs me in.

Reply:
I think the explore feed might be showing before the nav stack's restored because Explore is the first VC that's being restored. I took out the `dispatch_async` here https://github.com/wikimedia/wikipedia-ios/pull/3135/commits/bf97cd236b23123dc2d5aabbee0fe6712705f576 assuming that `resumeApp` will always be called from the main thread - is that what you were thinking?

Reply:
Settings can handle login ambiguity but there are 2 separate VCs for logged in / logged out state so it'd require merging those 2 into 1 VC

Reply:
I think merging them and not being dependent on that call would be ideal but might take more time

Reply:
That main queue async can remain as it doesn't affect the issue I'm seeing. What I'm seeing is the splash screen is removed, the stack restoration is kicked off, the stack restoration makes a network call to verify I'm logged in before restoring the stack, because of that async network call the explore feed shows since the splash screen was removed, then the whole stack is restored when the network call returns. When logged out, it seems to work as expected because there's no async network call to verify the user is logged in.

Reply:
If merging and updating the VCs would take too long, a quicker option would be to only attempt login when the account VC is on the stack and ensure the splash screen stays visible until the login attempt callback is called asynchronously. This would ensure most app launches won't be slowed down by the login call and also ensure the explore screen doesn't show briefly before the restored nav stack when the login call needs to be made